third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1221,4 +1515,257 @@ void schedulePersistentKernel(Fusion* fusion, const ReductionParams& rparams) {
+      // iteration domain
+      int axisID = -1;
+      if (rparams.vectorization_factor_tmp_gmem_read > 1) {
+        outer_reduction_tv->split(
+            axisID, rparams.vectorization_factor_tmp_gmem_read);
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::Unroll);
+      }
+      if (rparams.lparams.bdimx() > 1) {
+        outer_reduction_tv->split(
+            axisID, NamedScalar::getParallelDim(ParallelType::TIDx));
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::TIDx);
+      }
+
+      outer_reduction_tv->split(
+          axisID, NamedScalar::getParallelDim(ParallelType::BIDy));
+      outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::BIDy);
+    }
+
+    auto outer_reference_tv =
+        reduction_scheduler_utils::sortAndRFactor(outer_reduction_tv);
+    outer_reference_tvs.emplace_back(outer_reference_tv);
+  }
+
+  for (auto output : dummy_outputs) {
+    fusion->addOutput(output);
+  }
+
+  const bool vectorize =
+      rparams.vectorize_inner_reduction || rparams.vectorize_iter_dom;
+  const bool is_outer_grid_persistence = rparams.persistent_kernel &&
+      rparams.cross_grid_inner_reduction && !rparams.fastest_dim;
+
+  // Propagate inner reduction
+  reduction_scheduler_utils::propagateTransformation(
+      inner_reference_tv, boundaryNodesSet);
+  reduction_scheduler_utils::propagateRFactor(
+      inner_reference_tv, inner_reduction_tv, inner_reduction_tvs);
+  reduction_scheduler_utils::propagateParallelization(
+      fusion,
+      inner_reduction_tv,
+      inner_reference_tv,
+      unroll,
+      vectorize,
+      is_outer_grid_persistence,
+      inner_reduction_tvs,
+      cached_inputs,
+      cached_outputs);
+
+  // Propagate outer reduction
+  for (long unsigned int i = 0; i < outer_reference_tvs.size(); i++) {
+    reduction_scheduler_utils::propagateTransformation(
+        outer_reference_tvs[i], boundaryNodesSet);
+    reduction_scheduler_utils::propagateParallelization(
+        fusion,
+        outer_reduction_tvs[i],
+        outer_reference_tvs[i],
+        unroll,
+        vectorize,
+        is_outer_grid_persistence,
+        outer_reduction_tvs,
+        cached_inputs,
+        cached_outputs);
+  }
+
+  // special vectorization of temp gmem
+  if (rparams.vectorization_factor_tmp_gmem_write > 1) {
+    for (auto tv : cached_gmem) {
+      tv->split(-1, rparams.vectorization_factor_tmp_gmem_write);
+      tv->axis(-1)->parallelize(ParallelType::Vectorize);
+    }
+  }
+  if (rparams.vectorization_factor_tmp_gmem_read > 1) {
+    for (auto tv : cached_gmem_reload) {
+      tv->axis(-2)->parallelize(ParallelType::Vectorize);
+    }
+  }
+
+  if (rparams.vectorization_factor_tmp_gmem_read > 1) {
+    for (auto tv_pair : cached_outputs) {
+      if (tv_pair.second->axis(-1)->getParallelType() !=
+          ParallelType::Vectorize) {
+        tv_pair.second->axis(-1)->parallelize(ParallelType::Vectorize);
+      }
+    }
+  }
```

Comment:
IIUC, this means that the read of the temp buffer is vectorized, all of the outputs are vectorized. Why?

Reply:
added comments.
```
  // if the read of temp gmem is vectorized, each thread needs to process these
  // vectorized loaded items, then we can also vectorize the outputs of outer
  // reductions. Its defination also implies the output of inner reductions are
  // vectorized so we don't need to bother filtering them out.
```

Reply:
may be this is more easier to understand? `Since the output of inner reductions are already vectorized,
  // so we don't need to bother filtering them out in this for loop.` This for-loop just want to vectorize the `output of outer reduction`, however, it also checks `output of inner reduction`. But this additional check doesn't cause us any issue because `output of inner reductions are already vectorized`. So we don't need to test if the outer is from inner reduction of outer reduction.

The output domain of outer reudction is the inner domain of the input, e.g. [2018,8192], the output domain of outer reduction is 8192, which is ensured to be vectorizible by runtime check.

Reply:
> This for-loop just want to vectorize the output of outer reduction, however, it also checks output of inner reduction. But this additional check doesn't cause us any issue because output of inner reductions are already vectorized. 

OK, this makes sense, but it assumes all of the `cached_outputs` have an innermost domain that's mapped with the vectorized domain of the temp buffer read. What if a fusion also had an auxiliary output of some intermediate results? Maybe saving partial reduction results of the inner reduction. Again, I can't imagine why it should happen, but as far as it's allowed, it should work correctly.

I think here we should use `parallelizeAllLike` in a similar way as: https://github.com/NVIDIA/Fuser/blob/main/csrc/scheduler/reduction_utils.cpp#L460-L466

Reply:
Did you address this?